File: src/interpret/hw_01.go
package main

import (
	"fmt"
	"io/ioutil"
)

var table = [][]byte{
	{'b', 'e', 'g', 'i', 'n'},
	{'e', 'n', 'd'},
	{'i', 'n', 't', 'e', 'g', 'e', 'r'},
	{'r', 'e', 'a', 'l'},
	{'='},
	{'+'},
	{'+', '+'},
	{'*'},
	{','},
	{';'},
	{'('},
	{')'},
	{'#'},
}
var code = []byte{'{', '}', 'a', 'c', '=', '+', '$', '*', ',', ';', '(', ')', '#'}

func init() {

}

type code_val struct {
	code byte   //单词种别
	val  []byte //单词值
}

func reverse(token []byte) byte {
	length := len(token)
l0:
	for index := range table {
		if len(table[index]) != length {
			continue
		}
		i := 0
		for i = range table[index] {
			if table[index][i] != token[i] {
				continue l0
			}
		}
		if i == length-1 {
			return code[index]
		}
	}
	return 'i'
}

func scanner(buf []byte, i *int) code_val {
	codeVal := code_val{}
	token := make([]byte, 0)
	length := len(buf)
	if buf[*i] >= 'a' && buf[*i] <= 'z' {
		for *i < length && (buf[*i] >= 'a' && buf[*i] <= 'z' || buf[*i] >= '0' && buf[*i] <= '9') {
			token = append(token, buf[*i])
			*i++
		}
		codeVal.code = reverse(token)
		//TODO 此处我先式注释掉
		if codeVal.code == 'i' {
			codeVal.val = token
		}
		return codeVal
		//处理数字 123
	} else if buf[*i] >= '0' && buf[*i] <= '9' {
		for buf[*i] >= 'a' && buf[*i] <= 'z' || buf[*i] >= '0' && buf[*i] <= '9' {
			token = append(token, buf[*i])
			*i++
		}
		//处理小数 0.3
		if buf[*i] == '.' {
			token = append(token, buf[*i])
			*i++
			for buf[*i] >= '0' && buf[*i] <= '9' {
				token = append(token, buf[*i])
				*i++
			}
			codeVal.code = 'y'
		} else {
			codeVal.code = 'x'
		}
		codeVal.val = token
		return codeVal
	}
	//实数   .123
	if buf[*i] == '.' {
		token = append(token, buf[*i])
		*i++
		if buf[*i] >= '0' && buf[*i] <= '9' {
			for buf[*i] >= '0' && buf[*i] <= '9' {
				token = append(token, buf[*i])
				*i++
			}
			codeVal.code = 'y'
			codeVal.val = token
			return codeVal
		}
		panic("Error Word " + string(token))
	}
	switch buf[*i] {
	case ',':
		codeVal.code = ','
		break
	case ';':
		codeVal.code = ';'
		break
	case '(':
		codeVal.code = '('
		break
	case ')':
		codeVal.code = ')'
		break
	case '=':
		codeVal.code = '='
		break
	case '+':
		if buf[*i+1] == '+' {
			codeVal.code = '$'
			*i++
		} else {
			codeVal.code = '+'
		}
		break
	case '*':
		codeVal.code = '*'
		break
	case '#':
		codeVal.code = '#'
		break
	default:
		panic("Error character: " + string(buf[*i]))
	}
	*i++
	return codeVal
}

//删除注释以及\n\r
func preConstruct(filePath string) []byte {
	bytes, err := ioutil.ReadFile("F:\\IdeaProjects\\leetcode\\src\\interpret\\test.txt")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(bytes))
	stringFlag := false
	j := 0
	size := len(bytes)
	for i := 0; i < size; i++ {
		if i > 0 && bytes[i] == ' ' && bytes[i-1] == ' ' {
			continue
		}
		if bytes[i] == '"' {
			stringFlag = !stringFlag
		}
		if !stringFlag {
			//第一类注释 //xxxx\n 遇到换行符结束
			if i < size-1 && bytes[i] == '/' && bytes[i+1] == '/' {
				i++
				for bytes[i] != '\r' {
					i++
				}
				i++
			} else if i < size-3 && bytes[i+1] == '*' && bytes[i+2] == '*' && bytes[i] == '/' {
				//第三类注释 /**xxxx**/
				i = i + 3
				for bytes[i+2] != '/' && bytes[i+1] != '*' && bytes[i] != '*' {
					i++
				}
				i = i + 4
			} else if i < size-2 && bytes[i] == '/' && bytes[i+1] == '*' {
				//第二类注释，/*xxx*/
				i = i + 2
				for bytes[i] != '*' && bytes[i+1] != '/' {
					i++
				}
				i = i + 2
			}
		}
		bytes[j] = bytes[i]
		j++
	}
	for i := 0; i <= j; i++ {
		if bytes[i] == '\n' || bytes[i] == '\r' {
			bytes[i] = ' '
		}
	}
	k := 0
	for i := 0; i < j; i++ {
		if i > 0 && bytes[i] == ' ' && bytes[i-1] == ' ' {
			continue
		}
		bytes[k] = bytes[i]
		k++
	}
	bytes[k] = '#'
	if bytes[0] == ' ' {
		return bytes[1 : k+1]
	}
	return bytes[:k+1]
}

func main() {
	bytes := preConstruct("F:\\IdeaProjects\\leetcode\\src\\interpret\\test.txt")
	fmt.Println()
	fmt.Println(string(bytes))
	fmt.Println()
	doScan(bytes)
}

func doScan(buf []byte) {
	i := 0
	for true {
		for buf[i] == ' ' {
			i++
		}
		val := scanner(buf, &i)
		fmt.Println(string(val.code), string(val.val))
		if val.code == '#' {
			break
		}
	}
}

File: src/interpret/Hw_01.java
package interpret;

import java.io.IOException;

public class Hw_01 {
    private static final int len = 20;
    private static final String[] table = new String[]{"begin", "end", "integer", "real"};
    private static final char[] codes = new char[]{'{', '}', 'a', 'c'};

    static class CodeVal {
        public char code;//单词的类别
        public String val;//单词值

        public CodeVal(char code, String val) {
            this.code = code;
            this.val = val;
        }

        public CodeVal() {

        }
    }

    /**
     * 查基本字表函数
     *
     * @param token
     * @return
     */
    public char reverse(String token) {
        for (int i = 0; i < codes.length; i++) {
            if (token.equals(table[i])) {
                return codes[i];
            }
        }
        return 'i';
    }

    public CodeVal scanner(String buf, int i) {
        CodeVal node = new CodeVal();
        StringBuilder token = new StringBuilder();
        if (buf.charAt(i) >= 'a' && buf.charAt(i) <= 'z') {
            while (buf.charAt(i) >= 'a' && buf.charAt(i) <= 'z' || buf.charAt(i) >= '0' && buf.charAt(i) <= '9') {
                token.append(buf.charAt(i++));
            }
            String s = token.toString();
            node.code = reverse(s);
            if (node.code == 'i') {
                node.val = s;
            }
            return node;
        } else if (buf.charAt(i) >= '0' && buf.charAt(i) <= '9') {
            while (buf.charAt(i) >= '0' && buf.charAt(i) <= '9') {
                token.append(buf.charAt(i++));
            }
            if (buf.charAt(i) == '.') {
                token.append(buf.charAt(i++));
                while (buf.charAt(i) >= '0' && buf.charAt(i) <= '9') {
                    token.append(buf.charAt(i++));
                }
                node.code = 'y';
            } else {
                node.code = 'x';
            }
            String s = token.toString();
            node.val = s;
        } else if (buf.charAt(i) == '.') {
            token.append(buf.charAt(i++));
            if (buf.charAt(i) >= '0' && buf.charAt(i) <= '9') {
                while (buf.charAt(i) >= '0' && buf.charAt(i) <= '9') {
                    token.append(buf.charAt(i++));
                }
                node.code = 'y';
                node.val = token.toString();
            } else {
                System.out.println("Error word->" + token);
                throw new IllegalArgumentException(token.toString());
            }
        }
        switch (buf.charAt(i)) {
            case ',': {
                node.code = ',';
                break;
            }
            case ';': {
                node.code = ';';
                break;
            }
            case '(': {
                node.code = '(';
                break;
            }
            case ')': {
                node.code = ')';
                break;
            }
            case '=': {
                node.code = '=';
                break;
            }
            case '+': {
                if (buf.charAt(++i) == '+') {
                    node.code = '$';
                } else {
                    node.code = '+';
                    i--;
                }
                break;
            }
            case '*': {
                node.code = '*';
                break;
            }
            case '#': {
                node.code = '#';
                break;
            }
            default: {
                throw new IllegalArgumentException(i + ":" + buf.charAt(i));
            }
        }
        i++;
        return node;

    }

    public static void main(String[] args) throws IOException, InterruptedException {
//        String[] strs = new String[]{"A", "B", "C", "D"};
//        for (int i = 0; i < 6; i++) {
//            for (int j = 0; j < strs.length; j++) {
//                int finalI = j;
//                Thread thread = new Thread(() -> {
//                    System.out.print(strs[finalI] + finalI + " ");
//                });
//                thread.start();
//                thread.join();
//            }
//            Thread.sleep(100);
//            System.out.println();
//        }
    }
}
